### ISSUE TITLE
YOLOv11 Model not fully quantized to int8

### ISSUE THREAD

**Issue by ssdv1** [ultralytics/ultralytics#17441]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Export

### Bug

the output file is not fully int8. It has some float32
[best_int8.zip](https://github.com/user-attachments/files/17688543/best_int8.zip)

### Environment

Ultralytics 8.3.28  Python-3.8.20 torch-2.4.1+cpu CPU (Intel Core(TM) i5-8250U 1.60GHz)
Setup complete  (8 CPUs, 11.9 GB RAM, 160.5/222.9 GB disk)

### Minimal Reproducible Example

from ultralytics import YOLO
model=YOLO("best.pt")
model.export(format="tflite", int8=True, data="data.yaml")

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @ssdv1, thank you for your interest in Ultralytics 🚀! We recommend taking a look at our [Docs](https://docs.ultralytics.com) where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples that might help address common concerns and functionalities.

Regarding your 🐛 Bug Report on the YOLOv11 model quantization, could you please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) if not already included? This will greatly assist us in diagnosing the issue.

To ensure your setup aligns with the latest updates, please upgrade to the latest `ultralytics` package along with all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

For more personalized help, you can engage with the Ultralytics community on various platforms. Check out our [Discord](https://ultralytics.com/discord) for real-time discussions or head over to [Discourse](https://community.ultralytics.com) or our [Subreddit](https://reddit.com/r/ultralytics) for in-depth conversations and community knowledge sharing.

## Environments

Please run YOLO in any of these verified environments to ensure compatibility:

- **Notebooks**: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud**: Follow the [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon**: Head to the [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker**: Check out our [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

Here’s the current status of our continuous integration:

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [CI tests](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) are passing successfully, verifying consistent performance across supported operating systems.

This issue response is automated, but an Ultralytics engineer will follow up soon to offer additional support. Thank you for your patience and contribution! 🌟

**Comment by Y-T-G** [comment]
Are you checking the `full_integer_quant` model?

**Comment by Y-T-G** [comment]
That's normal. The biases are only converted to int32.

https://github.com/tensorflow/model-optimization/issues/775#issuecomment-893104919

**Comment by Y-T-G** [comment]
https://stackoverflow.com/a/64487684/8061030

**Comment by Y-T-G** [comment]
It's dynamically quantized which means it will have a mix of FP32 and INT8

https://blog.tensorflow.org/2024/04/faster-dynamically-quantized-inference-with-xnnpack.html

> Dynamically quantized models are similar to fully-quantized models in that the weights for the Fully Connected and Convolution operators are quantized to 8-bit integers during model conversion. All other tensors are not quantized, they remain as float32 tensors.

**Comment by Y-T-G** [comment]
> > https://stackoverflow.com/a/64487684/8061030
> 
> why is the model larger in size(28mb)

In your screenshot, it isn't. It's 2.2MB.

**Comment by Y-T-G** [comment]
Something wrong with your disk storage then if the uncompressed size is that much larger.

```python

In [14]: !ls -sh yolo11n_saved_model/yolo11n_full_integer_quant.tflite
3.2M yolo11n_saved_model/yolo11n_full_integer_quant.tflite

In [15]: !ls -sh yolo11n_saved_model/yolo11n_int8.tflite
2.9M yolo11n_saved_model/yolo11n_int8.tflite
```

**Comment by Y-T-G** [comment]
Did you try passing your own data.yaml for calibration during export?

**Comment by Y-T-G** [comment]
Can you try using "coco128.yaml" for export?

**Comment by Y-T-G** [comment]
I guess then provide create a test set with only 300 images from your datasey and then use that yaml.

**Comment by glenn-jocher** [comment]
It seems the larger size of the fully quantized model might be due to additional metadata or calibration-related data. To minimize size, ensure you are using a minimal and representative dataset for calibration. If the issue persists, consider consulting the [YOLO Common Issues guide](https://docs.ultralytics.com/guides/yolo-common-issues/) or reach out for further troubleshooting.

**Comment by glenn-jocher** [comment]
It seems like the calibration process might be sensitive to the size of the dataset used. When using exactly 128 images, it aligns with the default sample size often used for quantization calibration. If the model size is appropriate with this, it’s likely the calibration process is optimized for this count. You may continue using this setup for better results.

**Comment by Y-T-G** [comment]
It should be fixed in the latest onnx2tf version. Run `pip install -U onnx2tf` and export.

https://github.com/PINTO0309/onnx2tf/issues/724

**Comment by Y-T-G** [comment]
`tensorflow==2.17` should fix the issue with file size.

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has been merged into the `main` branch via [PR #18467](https://github.com/ultralytics/ultralytics/pull/18467). 🎉  

### Summary of Changes  
Key updates in the PR include:  
- **Updated `onnx2tf` Version Range**: The upper version limit of the `onnx2tf` library has been increased from `1.22.3` to `1.26.3`.  
  - 📈 This update enhances compatibility by allowing the export pipeline to support newer versions of `onnx2tf`.  
  - 🛠️ It also improves stability and addresses potential issues or bugs resolved in recent library updates.  
These changes aim to ensure a smoother export process for users working with TensorFlow workflows. 🚀  

Special thanks to the PR author [@Y-T-G](https://github.com/Y-T-G) and contributor [@Laughing-q](https://github.com/Laughing-q) for their efforts on this! 🙌  

### Testing the Fix  
To test whether these changes resolve the issue, you can:  
1. Install the latest updates directly from the `main` branch using:  
   ```bash  
   pip install git+https://github.com/ultralytics/ultralytics.git@main  
   ```  
2. Alternatively, wait for the next official PyPI release, which will include these changes.  

### Request for Feedback  
Please test these changes and let us know if they resolve the issue for you. Your feedback is incredibly valuable in ensuring everything works as expected.  

Thank you so much for reporting this issue and helping us improve! 🙏 If the problem persists, don’t hesitate to share additional details or feedback here.

**Comment by pimakshay** [comment]
> It seems like the calibration process might be sensitive to the size of the dataset used. When using exactly 128 images, it aligns with the default sample size often used for quantization calibration. If the model size is appropriate with this, it’s likely the calibration process is optimized for this count. You may continue using this setup for better results.

@glenn-jocher if the integer quant model (best_integer_quant.tflite) size is larger than the best_float32.tflite model, will it still be faster than a float 32 model? Will the performance of the integer quant model be better on Android phones (considering TFLite is most suitable for Androids)?

**Comment by glenn-jocher** [comment]
@pimakshay Yes, the integer quantized model will still typically be faster than the float32 model during inference, even if the file size is larger. This is because:

1. Integer operations are computationally more efficient than floating-point operations on most hardware
2. The larger file size is often due to quantization metadata and calibration tables, not computational overhead

Regarding Android performance: TFLite models with INT8 quantization are specifically optimized for mobile devices and generally perform significantly better on Android phones compared to desktop environments. Most modern Android devices have hardware acceleration for INT8 operations through the Neural Networks API and vendor-specific accelerators (like Qualcomm's Hexagon DSP), which can provide 2-4x faster inference than float32.

For best results on Android, you might want to experiment with different export configurations to find the optimal balance between model size, inference speed, and accuracy for your specific use case.

**Comment by axion66** [comment]
Hi!

For my case, setting `imgsz` that's smaller than image size used to train helped me the model to quantize into int8.

from
`yolo export model=final.pt format=tflite int8=True data="data.yaml"` (trained imgsz was 640 by 640)
to
`yolo export model=final.pt format=tflite int8=True data="data.yaml" imgsz="320,320"`

**Comment by glenn-jocher** [comment]
Nice find—TFLite INT8 PTQ can be shape-sensitive; exporting and calibrating at a smaller, fixed `imgsz` often keeps more ops quantized and reduces size. To make this reproducible, export with the exact `imgsz` you’ll deploy and calibrate on that size (you can limit calibration with `fraction`); see the [TFLite export arguments](https://docs.ultralytics.com/integrations/tflite/#export-arguments).

```bash
yolo export model=final.pt format=tflite int8=True data=data.yaml imgsz=320 fraction=0.5
```
